### ISSUE TITLE
Slow matrix multiplication

### ISSUE THREAD

**Issue by warnerjon12** [sympy/sympy#15421]
Matrix multiplication currently uses the naive algorithm. It would be good to implement a faster one for large matrices.

**Comment by bjodah** [comment]
@warnerjon12 what do you mean by "naive algorithm"?

**Comment by warnerjon12** [comment]
I had thought there were significantly better algorithms, but I guess the Strassen algorithm is the best known. Still it would be good to implement this after some cutoff point.

**Comment by bjodah** [comment]
If we can show measurable performance benefit of an alternative algorithm, then that's great (though I suspect matrices will be too big before that happens).

**Comment by Mohitbalwani26** [comment]
i would like to work on the issue so can u please guide me?

**Comment by warnerjon12** [comment]
I don't know how much speedup we will get in Python (it may only be better
for fairly large matrices), but start here:
https://en.wikipedia.org/wiki/Strassen_algorithm#Algorithm

**Comment by siefkenj** [comment]
A very naive implementation gives a speedup for sizes >= 128x128. This might be able to be optimized down to 32x32, but probably beyond that the overhead of the extra lists will overwhelm the advantage.

```
def strassen(mat, other):
    n = mat.shape[0]

    a11, a12, a21, a22 = mat[:n//2,:n//2], mat[:n//2,n//2:], mat[n//2:,:n//2], mat[n//2:,n//2:]
    b11, b12, b21, b22 = other[:n//2,:n//2], other[:n//2,n//2:], other[n//2:,:n//2], other[n//2:,n//2:]
    
    m1 = (a11+a22)*(b11+b22)
    m2 = (a21+a22)*b11
    m3 = a11*(b12-b22)
    m4 = a22*(b21-b11)
    m5 = (a11+a12)*b22
    m6 = (a21-a11)*(b11+b12)
    m7 = (a12-a22)*(b21+b22)
    
    c11 = m1+m4-m5+m7
    c12 = m3+m5
    c21 = m2+m4
    c22 = m1-m2+m3+m6
    return c11.hstack(c11.vstack(c11,c21), c11.vstack(c12,c22))
```

**Comment by warnerjon12** [comment]
Go ahead and push something, it only adds one extra check for smaller
matrices (size check). However, you will need to look at the overhead of
padding to a power of two size.

**Comment by rishabh-bhatnagar** [comment]
Is this issue resolved or may I try resolving it??

**Comment by warnerjon12** [comment]
Honestly I don't know how much it is worth. There is only so much to be
done in Python. If you wanted to resolve it, you would probably implement
the Strassen algorithm in C using the Python C API and find a good
switching point. Otherwise it seems to be the case that it doesn't make
much difference in Python.

**Comment by asmeurer** [comment]
For a C implementation you should look at implementing it in symengine, though I'm not sure if symengine has matrices yet.

**Comment by rishabh-bhatnagar** [comment]
Yes. I'll be taking up the issue and try to optimise multiplication.

**Comment by bjodah** [comment]
It does: https://github.com/symengine/symengine/blob/05a7f99811f5ebd2b0ffb2acf69840447952a541/symengine/dense_matrix.cpp#L342

**Comment by sylee957** [comment]
There are more than Strassen algorithm, and I also see that there is Laderman's algorithm that can multiply 3*3 matrix.

https://www.ams.org/journals/bull/1976-82-01/S0002-9904-1976-13988-2/S0002-9904-1976-13988-2.pdf
https://arxiv.org/pdf/1508.01110.pdf
